yt_dlp/extractor/youtube/pot/_director.py

```diff
@@ -0,0 +1,482 @@
+        settings = traverse_obj(
+            ie._downloader.params,
+            ('extractor_args', f'{EXTRACTOR_ARG_PREFIX}-{provider.PROVIDER_KEY.lower()}'))
+        logger = YoutubeIEContentProviderLogger(
+            ie, f'pot:{provider.PROVIDER_NAME}', log_level=log_level)
+        director.register_provider(provider(ie, logger, settings or {}))
+
+    for preference in _ptp_preferences.value:
+        director.register_preference(preference)
+
+    if director.logger.log_level <= director.logger.LogLevel.DEBUG:
+        # calling is_available() for every PO Token provider upfront may have some overhead
+        director.logger.debug(f'PO Token Providers: {provider_display_list(director.providers.values())}')
+        director.logger.debug(f'PO Token Cache Providers: {provider_display_list(cache.cache_providers.values())}')
+        director.logger.debug(f'PO Token Cache Spec Providers: {provider_display_list(cache.cache_spec_providers.values())}')
+        director.logger.trace(f'Registered {len(director.preferences)} provider preferences')
+        director.logger.trace(f'Registered {len(cache.cache_provider_preferences)} cache provider preferences')
+
+    return director
+
+
+def provider_display_list(providers: Iterable[IEContentProvider]):
+    def provider_display_name(provider):
+        display_str = join_nonempty(
+            provider.PROVIDER_NAME,
+            provider.PROVIDER_VERSION if not isinstance(provider, BuiltInIEContentProvider) else None)
+        statuses = []
+        if not isinstance(provider, BuiltInIEContentProvider):
+            statuses.append('external')
+        if not provider.is_available():
+            statuses.append('unavailable')
+        if statuses:
+            display_str += f' ({", ".join(statuses)})'
+        return display_str
+
+    return ', '.join(provider_display_name(provider) for provider in providers) or 'none'
+
+
+def clean_pot(po_token: str):
+    # Clean and validate the PO Token. This will strip invalid characters off
+    # (e.g. additional url params the user may accidentally include)
+    try:
+        return base64.urlsafe_b64encode(
+            base64.urlsafe_b64decode(urllib.parse.unquote(po_token))).decode()
+    except (binascii.Error, ValueError):
+        raise ValueError('Invalid PO Token')
+
+
+def validate_response(response: PoTokenResponse | None):
+    if (
+        not isinstance(response, PoTokenResponse)
+        or not response.po_token
+        or not isinstance(response.po_token, str)
+    ):  # noqa: SIM103
+        return False
+
+    try:
+        clean_pot(response.po_token)
+    except ValueError:
+        return False
+
+    return (
+        response.expires_at is None
+        or (
+            isinstance(response.expires_at, int)
+            and (
+                response.expires_at <= 0
+                or response.expires_at > int(dt.datetime.now(dt.timezone.utc).timestamp())
+            )
+        )
+    )
```

Comment:
yucky formatting. I mean it's fine but for the record I hate it

Reply:
yeah I'm not a fan of it too

Reply:
what about this ?
```suggestion
    if not isinstance(response.expires_at, int):
        return response.expires_at is None
    return response.expires_at <= 0 or response.expires_at > int(dt.datetime.now(dt.timezone.utc).timestamp())
```

&nbsp;

or we could make it a bit less verbose with `time.time()`
```suggestion
    if not isinstance(response.expires_at, int):
        return response.expires_at is None
    return response.expires_at <= 0 or response.expires_at > int(time.time())
```